sxbp/sxbp_internal_get_file_size.c

```diff
@@ -0,0 +1,140 @@
+    struct stat file_info = { 0 };
+    if (fstat(file_descriptor, &file_info) == 0) {
+        // file size is file_info.st_size
+        *file_size = (size_t)file_info.st_size;
+        // return success
+        return SXBP_RESULT_OK;
+    } else {
+        // return error code
+        return SXBP_RESULT_FAIL_IO;
+    }
+}
+
+// Windows version
+#elif defined(_WIN32)
+// We're gonna need GetFileSizeEx() for this!
+#include <Windows.h>
+
+
+sxbp_result_t sxbp_get_file_size(FILE* file_handle, size_t* file_size) {
+    // preconditional assertions
+    assert(file_handle != NULL);
+    assert(file_size != NULL);
+    // try and convert the file handle to a Windows file descriptor
+    int file_descriptor = _fileno(file_handle);
+    if (file_descriptor < 0) {
+        // Windows API returns a negative code if it couldn't do the operation
+        return SXBP_RESULT_FAIL_IO;
+    }
+    // try and convert the Windows file descriptor to a Windows HANDLE
+    intptr_t windows_file_handle = _get_osfhandle(file_descriptor);
+    if (windows_file_handle == INVALID_HANDLE_VALUE) {
+        // Windows couldn't do it, return an error
+        return SXBP_RESULT_FAIL_IO;
+    }
+    // the file's size from the Windows API call will be stored in this struct
+    LARGE_INTEGER file_size_info;
+    // call the Windows API to check the file size
+    if (!GetFileSizeEx((HANDLE)windows_file_handle, &file_size_info)) {
+        // if it failed, return an error code
+        return SXBP_RESULT_FAIL_IO;
+    }
+    /*
+     * LARGE_INTEGER is actually a struct of two DWORDS (32-bits) with an extra
+     * member which is 64 bits on 64-bit systems.
+     * As it happens, we don't care about files that are larger than 1GiB so if
+     * the higher DWORD is set, we can return an error code instead as we don't
+     * need it (can't use it for input to the algorithm)
+     */
+    if (file_size_info.HighPart != 0) {
+        return SXBP_RESULT_FAIL_UNIMPLEMENTED;
+    } else {
+        *file_size = (size_t)file_size_info.LowPart;
+        return SXBP_RESULT_OK;
+    }
+}
+
+// Generic version
+#else
+sxbp_result_t sxbp_get_file_size(FILE* file_handle, size_t* file_size) {
+    // preconditional assertions
+    assert(file_handle != NULL);
+    assert(file_size != NULL);
+    /*
+     * seek to end
+     * NOTE: This isn't portable due to lack of meaningful support of `SEEK_END`
+     */
+    if (fseek(file_handle, 0, SEEK_END) != 0) {
+        // handle fseek() error
+        return SXBP_RESULT_FAIL_IO;
+    }
+    // get size and check it, before storing it or throwing an error
+    int check_file_size = ftell(file_handle);
+    if (check_file_size < 0) {
+        // handle ftell() error
+        return SXBP_RESULT_FAIL_IO;
+    }
+    *file_size = (size_t)check_file_size;
+    // seek to start again
+    fseek(file_handle, 0, SEEK_SET);
+    // return the calculated file size
+    return SXBP_RESULT_OK;
+}
+#endif
```

Comment:
I don't like this generic version. It does things in a fundamentally different way to the others. I think it should be removed instead.